lucene/benchmark/src/java/org/apache/lucene/benchmark/byTask/utils/StreamUtils.java

```diff
@@ -0,0 +1,113 @@
+
+  /** File format type */
+  public enum Type {
+    /** BZIP2 is automatically used for <b>.bz2</b> and <b>.bzip2</b> extensions. */
+    BZIP2(CompressorStreamFactory.BZIP2),
+    /** GZIP is automatically used for <b>.gz</b> and <b>.gzip</b> extensions. */
+    GZIP(CompressorStreamFactory.GZIP),
+    /** Plain text is used for anything which is not GZIP or BZIP. */
+    PLAIN(null);
+    private final String csfType;
+    Type(String csfType) {
+      this.csfType = csfType;
+    }
+    private InputStream inputStream(InputStream in) throws IOException {
+      try {
+        return csfType==null ? in : new CompressorStreamFactory().createCompressorInputStream(csfType, in);
+      } catch (CompressorException e) {
+        throw new IOException(e.getMessage(), e);
+      }
+    }
+    private OutputStream outputStream(OutputStream os) throws IOException {
+      try {
+        return csfType==null ? os : new CompressorStreamFactory().createCompressorOutputStream(csfType, os);
+      } catch (CompressorException e) {
+        throw new IOException(e.getMessage(), e);
+      }
+    }
+  }
+
+  private static final Map<String,Type> extensionToType = new HashMap<>();
+  static {
+    // these in are lower case, we will lower case at the test as well
+    extensionToType.put(".bz2", Type.BZIP2);
+    extensionToType.put(".bzip", Type.BZIP2);
+    extensionToType.put(".gz", Type.GZIP);
+    extensionToType.put(".gzip", Type.GZIP);
+  }
+  
+  /**
+   * Returns an {@link InputStream} over the requested file. This method
+   * attempts to identify the appropriate {@link InputStream} instance to return
+   * based on the file name (e.g., if it ends with .bz2 or .bzip, return a
+   * 'bzip' {@link InputStream}).
+   */
+  public static InputStream inputStream(Path file) throws IOException {
+    // First, create a FileInputStream, as this will be required by all types.
+    // Wrap with BufferedInputStream for better performance
+    InputStream in = new BufferedInputStream(Files.newInputStream(file), BUFFER_SIZE);
+    return fileType(file).inputStream(in);
+  }
+
+  /** Return the type of the file, or null if unknown */
+  private static Type fileType(Path file) {
+    Type type = null;
+    String fileName = file.getFileName().toString();
+    int idx = fileName.lastIndexOf('.');
+    if (idx != -1) {
+      type = extensionToType.get(fileName.substring(idx).toLowerCase(Locale.ROOT));
+    }
+    return type==null ? Type.PLAIN : type;
+  }
+  
+  /**
+   * Returns an {@link OutputStream} over the requested file, identifying
+   * the appropriate {@link OutputStream} instance similar to {@link #inputStream(Path)}.
+   */
+  public static OutputStream outputStream(Path file) throws IOException {
+    // First, create a FileInputStream, as this will be required by all types.
+    // Wrap with BufferedInputStream for better performance
+    OutputStream os = new BufferedOutputStream(Files.newOutputStream(file), BUFFER_SIZE);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *os*. Currently, there are execution paths that do not contain closure statements, e.g., when exception is thrown by Type.outputStream. Either a) close *os* in a try-finally block or b) close the resource by declaring *os* in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).